### PR TITLE
fix(price-oracle): pool-implied ground-truth re-anchor for poisoned anchors (closes #784, #785)

### LINKED ISSUE #784: [price-oracle] Downward glitch poisons anchor → upward-only spike guard freezes correct heals for 14 days (YGG)
## Symptom / Evidence

A single bad downward read becomes the price anchor, after which the upward-only spike guard rejects **every correct recovery** for the full `PRICE_SPIKE_STALENESS_MS` window (14 days).

- YGG (Base, **whitelisted**, `0xaAC78d1219c08AecC8e37e03858FE885f5EF1799`, in WETH/YGG CL pool `0xdC9279CE30A5044bbc345bf50030d44f45f0e768`).
- One bad read priced YGG at **$0.002596** at block ~36679673 (2025-10-11); true price ~**$0.14** throughout (verified on-chain via slot0; 2nd-lowest price ever was $0.111357, ~43× above the glitch).
- Deployed snapshots: continuous hourly snapshots up to 2025-10-11 02:11 ($0.002596), then a **14-day gap with no accepted writes**, resuming 2025-10-25 02:37 at **$0.141**.
- Impact: YGG-side TVL and min-picked volume/fees understated ~54× for 14 days.

## Root cause

The spike guard (#730) is **upward-only** and protects the *anchor* — it has no defense against a **bad anchor**. Because the glitch was downward (downward moves are unguarded), it was accepted; every subsequent correct read then looked like a ≥10× upward spike and was rejected until the bad anchor aged out of the 14-day staleness window.

There is no clean deterministic in-indexer rule that distinguishes a real heal from a spike on a single observation — only persistence over time or a second independent source separates them. This is the HITL design question.

## Fix (design options to evaluate)

- A two-sided / second-source confirmation before an anchor is trusted (e.g. require N consistent reads, or confirm against a canonical baseline).
- A bounded re-anchor when a long run of consistent "spikes" all agree (the run is evidence the anchor, not the reads, is wrong).
- Shorter staleness window with explicit trade-off analysis.

## Acceptance criteria

- [ ] A downward-glitch anchor no longer freezes correct heals for 14 days.
- [ ] The chosen rule is documented (including why a fully general deterministic rule is not possible).
- [ ] Regression test reproduces the YGG sequence and asserts recovery within a bounded number of reads.

## Blocked by

None — can start immediately (HITL: needs an architectural decision).

## Cross-references

#730, #695, #694, #668, #669

### LINKED ISSUE #785: [price-oracle] Low first-read becomes a permanent anchor — FIRST_FETCH_CAP guards only high reads (LIFE)
## Symptom / Evidence

A freshly-launched token's first oracle read, if too **low**, becomes a permanent poison anchor — every correct (higher) read afterward is rejected as a ≥10× spike, forever.

- LIFE/Cyberlife (Base, **non-whitelisted**, `0x151ccc74972a6f85690AfE997d0FDD69eFDC71FE`).
- First-ever read **$0.001442** at block ~37139400 (2025-10-21 17:35); the next read one hour later and all subsequent reads were **$0.22–$0.31** (true ~$0.25).
- The token has exactly **one** price snapshot — it got stuck on read #1.

## Root cause

`FIRST_FETCH_CAP` (#728) caps only a too-**high** first read (≤$10K for non-BTC). A too-**low** first read has no guard, so it is accepted as the anchor; from then on every correct read looks like an upward spike and is rejected.

Non-whitelisted, so trust-gated out of USD aggregates — the per-token price is wrong but pool TVL is not corrupted. Still worth fixing for price-table correctness and because a whitelisted token could hit the same path.

## Fix (design options to evaluate)

- Extend the first-fetch guard to a **low-side** bound as well (symmetric cap), or
- Bound the first read against a canonical baseline (the #700 direction), or
- Treat the first N reads as provisional until they agree.

## Acceptance criteria

- [ ] A too-low first read no longer becomes a permanent anchor.
- [ ] LIFE-style sequence recovers to the true price band within a bounded number of reads.
- [ ] Test covers both a too-high and a too-low first read.

## Blocked by

None — can start immediately (HITL: needs a decision on the low-side bound).

## Cross-references

#728, #700

### PR DESCRIPTION
**Pull request by FonsSicca**
## What

Fixes two price-oracle **poison-anchor** bugs with one mechanism — a pool-implied ground-truth check.

- **#784 (YGG, Base, whitelisted):** a downward glitch ($0.0026 vs true ~$0.14) poisoned the anchor; the upward-only spike guard (#730) then rejected every correct heal for the full 14-day staleness window.
- **#785 (LIFE, Base, non-whitelisted):** a too-low first read ($0.0014 vs true ~$0.25) became a *permanent* anchor — `FIRST_FETCH_CAP` (#728) only guards too-**high** first reads.

Both are the same failure: a bad-low value the read stream alone cannot disqualify, hitting the oracle at two different points.

## How

`refreshTokenPrice` gains an optional `impliedPriceHint` — the counterparty leg's trusted price × the pool's on-chain ratio, via new `getPoolImpliedUSD` (in `PriceTrust.ts`). It is an **independent witness** to the per-token oracle route: it comes from pool reserves / `sqrtPriceX96` (the stored `token0Price`/`token1Price` from #783), not from the route that froze/glitched.

- Computed in `loadPoolData` from the already-loaded pool + token entities (no new reads).
- Gated on the **counterparty** being trusted (whitelisted, non-blacklisted) — *not* the priced token — so non-whitelisted LIFE is still covered via its trusted WETH leg.
- **First-fetch (#785):** a too-far first read (too-low *or* too-high) is replaced by the hint → recovers in one read. Supersedes `FIRST_FETCH_CAP` when a trusted hint exists; a transient $0 read is left to the existing retry path.
- **Spike-rejection (#784):** a stuck-low anchor is re-anchored to the incoming read only when the read corroborates the hint **and** the anchor contradicts it.
- **Inert without a trusted counterparty:** the 3 reward-token `refreshTokenPrice` call sites and any pool lacking a trusted leg behave exactly as before.

### Decision flow

Every read walks these checks **in order** — first match wins. `band(a,b)` = "a and b are within 10× of each other, both nonzero" (`withinRatioBand`).

```
getTokenPrice (RPC) ─► read
  │
  ├─ read > $1M ? ─────────────────────────────────► REJECT  (ceiling #788 — hint ignored)
  │
  ├─ anchor == 0   (first fetch)
  │    ├─ hint present & NOT band(read, hint) ? ────► ANCHOR TO HINT      (#785)
  │    ├─ else read > $10k & not BTC ? ─────────────► REJECT to 0         (#728)
  │    └─ else ─────────────────────────────────────► accept read
  │
  ├─ read ≥ 10× anchor & anchor younger than 14d   (upward spike)
  │    ├─ band(read, hint) & NOT band(anchor, hint)?► RE-ANCHOR to read   (#784)
  │    └─ else ─────────────────────────────────────► REJECT, keep anchor (#730)
  │
  ├─ read == 0 & anchor fresh ? ───────────────────► keep anchor (last-known #694/#775)
  │
  └─ else ──────────────────────────────────────────► accept read (normal write)
```

### Worked examples

Canonical numbers (YGG ~$0.14, LIFE ~$0.25). ✅ = fix in action · 🛡 = guard holds (no bad write) · ➖ = behavior unchanged vs pre-PR · ⚠️ = known boundary.

| # | Situation | Anchor | RPC read | Hint | Branch that fires | Outcome |
|---|-----------|-------:|---------:|-----:|-------------------|---------|
| 1 | **#785** too-low first read (LIFE/WETH) | `0` | `$0.0014` | `$0.25` | pool-implied first-fetch | ✅ anchors to **$0.25** in one read |
| 2 | **#784** stuck-low anchor (YGG/WETH) | `$0.0026` | `$0.14` | `$0.14` | spike → re-anchor | ✅ accepts **$0.14**, breaks 14-day freeze |
| 3 | glitch-high read, **pool stays low** | `$0.14` | `$10,000` | `$0.15` | spike, `read` disagrees hint | 🛡 reject, keep **$0.14** *(the #801 audit killer)* |
| 4 | too-high first read, good hint | `0` | `$50,000` | `$0.20` | pool-implied first-fetch | ✅ anchors to **$0.20** (beats old reject-to-0) |
| 5 | untrusted counterparty / reward token | any | any | — | none (hint = 0) | ➖ identical to pre-PR |
| 6 | too-high first read, **no** hint | `0` | `$50,000` | — | `FIRST_FETCH_CAP` | 🛡 reject to **0** (#728 intact) |
| 7 | genuine spike, **no** hint | `$1` | `$15` | — | spike | 🛡 reject, keep **$1** (#730 intact) |
| 8 | in-band first read | `0` | `$0.24` | `$0.25` | none (falls through) | ✅ accepts **$0.24** — hint is a no-op |
| 9 | dead `$0` read | `$0.14` | `$0` | `$0.15` | last-known fallback | ➖ keep **$0.14** (hint never *manufactures* a price) |
| 10 | 10× jump but **anchor agrees hint** | `$0.10` | `$1.50` | `$0.12` | spike, `anchor` agrees hint | 🛡 reject, keep **$0.10** (anchor is the trusted one) |
| 11 | absurd read, corroborating hint | any | `$2,000,000` | `$0.15` | ceiling (#788) | 🛡 reject — ceiling is hint-proof |
| 12 | **glitched hint** at first fetch | `0` | — | `$5,000,000` | pool-implied first-fetch | ⚠️ anchors to **$5M** — see *Known boundary* |

### Can the hint overwrite a *correct* read?

Only in two narrow, self-bounding windows — and both require the **pool state itself** to be wrong, not just the oracle (a strictly higher bar than the single-glitch this fixes):

1. **First fetch, wrong hint** — a correct read >10× off a *bad* hint (manipulated CL spot at that block, or a stale counterparty price) anchors to the hint. Self-heals: a wrong-*high* mis-anchor is fixed by the next lower read (downward reads are unguarded); a wrong-*low* one recovers via the re-anchor path once the pool ratio normalises.
2. **Re-anchor, correlated double-glitch** — a correct anchor flipped only when a glitch-high read *and* a glitch-high hint agree with each other. Needs route + pool to fail high together, most plausible when the token's only liquidity is the hint pool itself.

A *single* glitch can never flip a correct price: the surviving correct source disagrees with the glitch and the 10× band catches it. Bounded further by the counterparty trust gate and the $1M read ceiling.

## Known boundary

The first-fetch branch anchors to the hint, which — unlike the incoming read — is not bounded by `MAX_ACCEPTED_PRICE` (#788) (row 12 above). In practice the hint = a trusted counterparty price (≤ $1M by its own ceiling) × the pool ratio, and a brand-new token implying >$1M would itself signal a glitched pool ratio. Left unguarded deliberately; happy to add a `hint ≤ MAX_ACCEPTED_PRICE` clamp if preferred.

## Test plan

- [x] `tsc --noEmit` — clean
- [x] `pnpm qa --write` — clean (biome)
- [x] `pnpm test` — **1481 passed, 33 skipped**
- [x] New `test/PriceOracle.test.ts` (6): #784 re-anchor; #785 too-low first read; #785 too-high first read (supersedes FIRST_FETCH_CAP); false-positive guard (#801 audit); #728 intact without a hint; inert without a hint
- [x] New `test/PriceTrust.test.ts` (3): `getPoolImpliedUSD` scaling + counterparty trust gate + zero/undefined guards

Closes #784, closes #785.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Enhanced token price handling with improved anchor recovery and first-fetch price determination using pool-derived pricing signals for greater accuracy.

* **Tests**
  * Added comprehensive test coverage for new price oracle scenarios including anchor recovery and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/velodrome-finance/indexer/pull/809?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by WooSungD — APPROVED** [review]

File: src/PriceOracle.ts [hunk 1]
```diff
@@ -284,6 +310,43 @@ export async function refreshTokenPrice(
...
+        healed.address,
+        chainId,
+        blockNumber,
+        new Date(blockTimestampMs),
+        hint,
+        healed.isWhitelisted,
+        context,
+      );
+      return anchored;
+    }
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply src/PriceOracle.ts:354]: Addressed in 63b0862.

Rather than clamping the hint to `MAX_ACCEPTED_PRICE`, I gated the first-fetch branch on `hint <= MAX_ACCEPTED_PRICE` so an out-of-ceiling hint is treated as *unusable* ground truth: the branch is skipped and the read takes the normal first-fetch path.

Reasoning for the deviation: clamping to $1M would still write an absurd anchor for, say, a sub-dollar token, and it diverges from the ceiling at line 301 which *rejects* rather than clamps. Skipping lets a reasonable read through and only degrades to pre-PR behavior in the glitched-hint case (which then self-heals via the re-anchor path once the ratio normalizes).

Covered by a new test: `#788: a hint above MAX_ACCEPTED_PRICE is not used as an anchor; the reasonable read is accepted instead`.
